engine/src/lib.rs

```diff
@@ -21,3 +21,299 @@ pub mod witness;
+
+use utilities::logging::ErrorType;
+
+pub fn settings_and_run_main(
+	settings_strings: Vec<String>,
+	start_from: state_chain_runtime::BlockNumber,
+) -> ExitStatus {
+	use_chainflip_account_id_encoding();
+	let opts = CommandLineOptions::parse_from(settings_strings);
+
+	let settings = match Settings::new_with_settings_dir(DEFAULT_SETTINGS_DIR, opts)
+		.context("Error reading settings")
+	{
+		Ok(settings) => settings,
+		Err(e) => {
+			eprintln!("Error reading settings: {}", e);
+			return ExitStatus { status_code: ERROR_READING_SETTINGS, at_block: NO_START_FROM };
+		},
+	};
+
+	match tokio::runtime::Builder::new_multi_thread()
+		.enable_all()
+		.build()
+		.unwrap()
+		.block_on(async {
+			// Note: the greeting should only be printed in normal mode (i.e. not for short-lived
+			// commands like `--version`), so we execute it only after the settings have been parsed.
+			utilities::print_start_and_end!(async run_main(settings, if start_from == NO_START_FROM { None } else { Some(start_from) }))
+		}) {
+		Ok(()) => ExitStatus { status_code: SUCCESS, at_block: NO_START_FROM },
+		Err(ErrorType::Error(e)) => {
+			if let Some(CreateStateChainClientError::CompatibilityError(block_compatibility)) =
+				e.downcast_ref::<CreateStateChainClientError>()
+			{
+				match block_compatibility.compatibility {
+					// we're no longer compatible, so we want to pass on the start to the one that is
+					// now compatible so that it can start from that number, ensuring we don't miss any blocks.
+					CfeCompatibility::NoLongerCompatible => ExitStatus {
+						status_code: engine_upgrade_utils::NO_LONGER_COMPATIBLE,
+						at_block: block_compatibility.at_block.number,
+					},
+					CfeCompatibility::NotYetCompatible => ExitStatus {
+						status_code: engine_upgrade_utils::NOT_YET_COMPATIBLE,
+						at_block: NO_START_FROM,
+					},
+					_ => {
+						unreachable!("We should never get here");
+					},
+				}
+			} else {
+				tracing::error!("Unknown error: {:?}", e);
+				ExitStatus { status_code: engine_upgrade_utils::UNKNOWN_ERROR, at_block: NO_START_FROM }
+			}
+		},
+		Err(ErrorType::Panic) =>
+			ExitStatus { status_code: engine_upgrade_utils::PANIC, at_block: NO_START_FROM },
+	}
+}
+
+async fn run_main(
+	settings: Settings,
+	start_from: Option<state_chain_runtime::BlockNumber>,
+) -> anyhow::Result<()> {
+	task_scope(|scope| {
+		async move {
+			let mut start_logger_server_fn =
+				Some(utilities::logging::init_json_logger(settings.logging.clone()).await);
+
+			let has_completed_initialising = Arc::new(AtomicBool::new(false));
+
+			let (state_chain_stream, unfinalised_state_chain_stream, state_chain_client) =
+				state_chain_observer::client::StateChainClient::connect_with_account(
+					scope,
+					&settings.state_chain.ws_endpoint,
+					&settings.state_chain.signing_key_file,
+					AccountRole::Validator,
+					true,
+					// TODO: remove wait for required version entirely
```

Comment:
I will do this in another PR. Because we're not running the engines side by side, we should be able to remove this now. (note this TODO: wasn't introduced here, the intention was to remove it once this PR is in)

Reply:
We don't need `wait_for_required_version` any more because we should never wait. The SCC should always fail out so that the runner can then start the required version.